### PR TITLE
Introduce get_anchor_credentials method as a replacement for lookup

### PR DESCRIPTION
**Pull request by frederikrothenberger**
`lookup` provides more information than necessary for a public endpoint.
This is why it is being deprecated. As a replacement `get_device_credentials` is introduced
which is more tailored to just provide the information required for authentication.

### CODE REVIEW

File: src/internet_identity_interface/src/lib.rs
Comment (przydatek) [inline src/internet_identity_interface/src/lib.rs:192]: I wonder whether a more specific/narrow name would be more appropriate here, as `Credential` is quite generic, and in a not-so-distant future we may add support for **verifiable credentials** (in the context of attribute sharing), so maybe we could avoid a potential future name collision.   IIUC, this is specific to a device, so maybe `DeviceCredential` would make sense?  (Note: if we change the struct name, then probably we'd also rename the related structs and functions, like `GetCredentialsResponse` or `get_credentials`). Wdyt?

File: src/internet_identity/src/storage/anchor.rs
Comment (przydatek) [inline src/internet_identity/src/storage/anchor.rs:170]: nit: maybe add a description of what happens when `filter` is `None`?
Alternatively, maybe drop `Option`, and leave it up to the caller to provide the dummy "catch-all" filter?  Maybe then also `Box` would be not needed?

File: src/internet_identity/src/storage/anchor.rs
Comment (przydatek) [reply src/internet_identity/src/storage/anchor.rs:89]: Why is it a `FnMut` rather than eg. `Fn(&Device) -> bool`?

**Review by przydatek — COMMENTED** [review]
Hi @frederikrothenberger, please see some questions comments below.
As this is my first review for II, please take the comments/suggestions with a grain of salt ;-)

File: src/internet_identity/src/storage/anchor.rs
Comment (frederikrothenberger) [reply src/internet_identity/src/storage/anchor.rs:89]: `filter` uses `FnMut` so I opted for the same type. But happy to change.

File: src/internet_identity/src/main.rs
Comment (frederikrothenberger) [reply src/internet_identity/src/main.rs:101]: I removed it for not being useful.

File: src/internet_identity/src/storage/anchor.rs
Comment (frederikrothenberger) [reply src/internet_identity/src/storage/anchor.rs:170]: Yeah, I think the whole interface gets a lot nicer if the `filter` is mandatory.

File: src/internet_identity_interface/src/lib.rs
Comment (frederikrothenberger) [reply src/internet_identity_interface/src/lib.rs]: I have renamed the interface. I don't mind the longer names. @nmattia wdyt?

**Review by przydatek — APPROVED** [review]
LGTM, thanks!
(Of course, let's wait for @nmattia's feedback)

File: src/internet_identity_interface/src/lib.rs
Comment (frederikrothenberger) [reply src/internet_identity_interface/src/lib.rs]: Having thought about this for a bit, I think `get_webauthn_credentials` would be most accurate, if not for the fact that we also have a bit of information in there about the existence recovery phrases. But at least it would be more specific than just `device`.

File: src/internet_identity/src/storage/anchor.rs
Comment (frederikrothenberger) [reply src/internet_identity/src/storage/anchor.rs:170]: If I only do `&Device` then I cannot do `.filter(filter)` but would have to first dereference `device`, i.e. something like `.filter(|&d| filter(d))`.

File: src/internet_identity/internet_identity.did
Comment (nmattia) [inline src/internet_identity/internet_identity.did:260]: I would suggest naming this `get_credentials` (or if really necessary `get_anchor_credentials`), I started looking at this assuming it would somehow only list the credentials for a device

File: src/internet_identity/internet_identity.did
Comment (nmattia) [inline src/internet_identity/internet_identity.did:218]: This is very descriptive, have you considered something like `Credentials` or `CredentialsInfo` or `AnchorCredentials`?

File: src/internet_identity/internet_identity.did
Comment (nmattia) [inline src/internet_identity/internet_identity.did:221]: nit: thinking about making things a bit more correct by construction. What happens if `{ recovery_credentials = [], recovery_phrase = true }`?

File: src/internet_identity/src/main.rs
Comment (nmattia) [inline src/internet_identity/src/main.rs:123]: Why not fetch all devices and partition them? If a new `Purpose` is introduced, we may forget to fetch it here

File: src/internet_identity/src/storage/anchor.rs
Comment (nmattia) [inline src/internet_identity/src/storage/anchor.rs:168]: I think it may be simpler to do everything in `get_query_credentials`, I find it a little odd that this module ("storage") needs to know about extracting credentials

File: src/internet_identity/tests/archive_integration_tests_pull.rs
Comment (nmattia) [inline src/internet_identity/tests/archive_integration_tests_pull.rs:265]: out of curiosity, why did this change?

File: src/internet_identity/internet_identity.did
Comment (nmattia) [reply src/internet_identity/internet_identity.did:221]: Err, I just realized there are no credentials for recovery _phrases_. So I'm guessing the recovery credentials are only for "recovery keys"?

Comment (frederikrothenberger) [reply src/internet_identity/internet_identity.did:221]: Yes, exactly. So your example would indicate no recovery devices but the existence of a recovery phrase.

File: src/internet_identity/internet_identity.did
Comment (frederikrothenberger) [reply src/internet_identity/internet_identity.did]: Makes sense. I have renamed it `get_anchor_credentials`.

File: src/internet_identity/src/main.rs
Comment (frederikrothenberger) [reply src/internet_identity/src/main.rs]: TIL about `partition`. Thanks for pointing it out.

File: src/internet_identity/src/storage/anchor.rs
Comment (frederikrothenberger) [reply src/internet_identity/src/storage/anchor.rs]: I originally had it that way, but without `partition` the code was very cumbersome. But since I now know it exists, I have removed this method again.

File: src/internet_identity/tests/archive_integration_tests_pull.rs
Comment (frederikrothenberger) [reply src/internet_identity/tests/archive_integration_tests_pull.rs:265]: I introduced a credential id on the `device_data_1()` test data (to make it a proper webauthn device), so the archive test that uses it now also archives the credential id.

File: src/internet_identity/internet_identity.did
Comment (nmattia) [inline src/internet_identity/internet_identity.did:221]: really nit: since we send the pubkeys everywhere, does it make sense to send it for the recovery phrase as well?

**Review by nmattia — APPROVED** [review]
Thanks a lot!

I didn't think that we may not be able to build a WebAuthn device since phrases don't have credentials. This is the ultimate nit so feel free to ignore, but couldn't we get rid of the `TryFrom` by folding over the devices once?

```rust
#[query]
fn get_foo(anchor_number: AnchorNumber) -> AnchorCredentials {
    let anchor = state::anchor(anchor_number);

    struct Foo { auth: Vec<WebauthnCredential>, recover: Vec<WebauthnCredential>, phrase: bool }

    let foo = anchor.into_devices().into_iter().fold(Foo { auth: vec![], recover: vec![], phrase: false}, |mut acc, d| {
        if d.key_type == KeyType::SeedPhrase {
            acc.phrase = true;
        } else if let Some(credential_id) = d.credential_id {
            if d.purpose == Purpose::Recovery {
                acc.recover.push(WebauthnCredential { pubkey: d.pubkey, credential_id })
            } else {
                acc.auth.push(WebauthnCredential { pubkey: d.pubkey, credential_id })
            }
        }
        acc
    });

    AnchorCredentials {
        credentials: foo.auth,
        recovery_credentials: foo.recover,
        recovery_phrase: foo.phrase,
    }
}
```

File: src/internet_identity/internet_identity.did
Comment (frederikrothenberger) [reply src/internet_identity/internet_identity.did]: I made it a `Vec<PubKey>` because there might still be anchors with multiple recovery phrases (from before we enforced the limit).

**Review by nmattia — APPROVED** [review]
Looking really nice, thanks!

**Pull request merged**
